### PR TITLE
[Woo POS] Remove ItemList view model

### PR DESCRIPTION
**Pull request by joshheald**
Part of: #14458

## Description

This PR removes the item list view model, in favour of treating the `ItemListView` as a viewmodel.

As mentioned in a previous PR, unfortunately using `EnvironmentObject` for injecting the posModel makes it impractical to write tests for the view(model) functions. That may also be true for AppStorage, I've not tried it yet though... we could try to address these shortcomings, but I haven't here.

If we did it, we would probably have to make private functions internal, so that they could be tested. I think that would be reasonable if the tests are valuable.

My view (🥁)  is that the code in the view is now simple enough that the tests we're missing wouldn't add value anyway... but what do you think, @staskus, @iamgabrielma, @jaclync? Is the `shouldShowHeaderBanner` logic too much to have in the view, untested? Should it be in the `posModel` instead?

Personally, I don't think so; it's a bit of view presentation logic to decide whether it's appropriate to show the banner, which is primarily a visual thing. The AppStorage macro isn't our code anyway, it's Apple's, so we simply change a bool on tapping the dismiss button.

## Testing information

It's worth testing the new way we use User Defaults; via AppStorage. I've checked this by dismissing the simple products banner on trunk, then checking it was still dismissed with this branch. 

The other changes relate to pagination, refresh, loading, and selecting items... but they're simply removing various ways of passing these calls through to the posModel, so exhaustive testing might not be the best use of time on this one. Item selection was the most complicated prior to this, the others were simple passthroughs from the view model.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/232">21.2</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by iamgabrielma** [comment]
>  using EnvironmentObject for injecting the posModel makes it impractical to write tests for the view(model) functions. That may also be true for AppStorage, I've not tried it yet though... we could try to address these shortcomings, but I haven't here.

Something that we might want to look into is mocking `POSAggregateModel`, declare the view in the test, inject the mock through `.environmentObject`, and use `ViewInspector` to see if all the buttons and other UI elements handled by changes in view state are behaving as they should. This is already part of our dependencies (context: pe5pgL-5Ee-p2) so nothing new we should be adding to the project.

File: WooCommerce/Classes/POS/Presentation/ItemListView.swift [hunk 1]
```diff
@@ -124,18 +130,18 @@ private extension ItemListView {
...
         ScrollView {
             VStack {
-                if dynamicTypeSize.isAccessibilitySize, viewModel.shouldShowHeaderBanner {
+                if dynamicTypeSize.isAccessibilitySize, shouldShowHeaderBanner {
                     bannerCardView
                 }
                 ForEach(items, id: \.productID) { item in
                     Button(action: {
-                        viewModel.select(item)
+                        posModel.addToCart(item)
```
Comment (iamgabrielma) [inline WooCommerce/Classes/POS/Presentation/ItemListView.swift:134]: It's nice to remove this extra layer of indirection, and just add items to cart directly from the view.

File: WooCommerce/Classes/POS/Presentation/ItemListView.swift [hunk 2]
```diff
@@ -253,11 +283,6 @@ private extension ItemListView {
 
 #if DEBUG
 #Preview {
-    ItemListView(
-        viewModel: ItemListViewModel(
-            posModel: PointOfSaleAggregateModel(
-                itemProvider: POSItemProviderPreview(),
-                cardPresentPaymentService: CardPresentPaymentPreviewService(),
-                orderService: POSOrderPreviewService())))
+    ItemListView()
```
Comment (iamgabrielma) [inline WooCommerce/Classes/POS/Presentation/ItemListView.swift:282]: 💯

**Comment by iamgabrielma** [comment]
> My view (🥁) is that the code in the view is now simple enough that the tests we're missing wouldn't add value anyway... but what do you think, @staskus, @iamgabrielma, @jaclync? Is the shouldShowHeaderBanner logic too much to have in the view, untested? Should it be in the posModel instead?

> Personally, I don't think so; it's a bit of view presentation logic to decide whether it's appropriate to show the banner, which is primarily a visual thing. The AppStorage macro isn't our code anyway, it's Apple's, so we simply change a bool on tapping the dismiss button.

Agree with this. We could make it internal temporarily while we make the rest of changes, or try the ViewInspector approach, but I'm not sure if is a good usage of our time. Adding to it, it would seem that with the new designs we won't have the one-time banner anymore, but we would have only the persistent toggle-able one, so I wouldn't spend much resources trying to work around this one specifically.

**Review by iamgabrielma — APPROVED** [review]
Thanks for the work on this Josh! Tests well, I haven't found any issue while smoke testing loading/adding/removing/clearing/reloading, etc, ...  🚢 🚢 🚢 🚢 🚢

**Comment by staskus** [comment]
> My view (🥁) is that the code in the view is now simple enough that the tests we're missing wouldn't add value anyway... but what do you think, @staskus, @iamgabrielma, @jaclync? Is the shouldShowHeaderBanner logic too much to have in the view, untested? Should it be in the posModel instead?

Yes, I think this is fine. If we really want to write tests, one option would be to inject aggregate model through the method parameter. func `shouldShowHeaderBanner(eligibleToShowSimpleProductsBanner: Bool) -> Bool`. As I mentioned before, ideally I would prefer not having POSModel within View Models at all. However, I don't know if it's always the most reasonable choice. Before making any decisions on this, I want to wait and see what remains out of view models in the end.

**Comment by joshheald** [comment]
> Something that we might want to look into is mocking `POSAggregateModel`, declare the view in the test, inject the mock through `.environmentObject`, 

Sadly, that's not possible; you have to use a concrete type with `EnvironmentObject`. This is because it has to conform to `ObservableObject`, and protocols can't fully do that.

> `ViewInspector` to see if all the buttons and other UI elements handled by changes in view state are behaving as they should. This is already part of our dependencies (context: pe5pgL-5Ee-p2) so nothing new we should be adding to the project.

I didn't know we already had it in the project, that's good to know 😊

**Pull request merged**